crates/apollo-compiler/src/ast/serialize.rs

```diff
@@ -831,6 +842,89 @@ fn serialize_string_value(state: &mut State, mut str: &str) -> fmt::Result {
     state.write("\"")
 }
 
+fn serialize_block_string(state: &mut State, str: &str) -> fmt::Result {
+    const TRIPLE_QUOTE: &str = "\"\"\"";
+    const ESCAPED_TRIPLE_QUOTE: &str = "\\\"\"\"";
+    const _: () = assert!(TRIPLE_QUOTE.len() == 3);
+    const _: () = assert!(ESCAPED_TRIPLE_QUOTE.len() == 4);
+
+    state.write(TRIPLE_QUOTE)?;
+    // `can_be_block_string` excludes \r, so the only remaining line terminator is \n
+    for mut line in str.split('\n') {
+        if line.is_empty() {
+            // Skip indentation which would be trailing whitespace
+            state.write("\n")?;
+            continue;
+        }
+        state.require_new_line()?;
+        loop {
+            if let Some((before, after)) = line.split_once(TRIPLE_QUOTE) {
+                state.write(before)?;
+                state.write(ESCAPED_TRIPLE_QUOTE)?;
+                line = after;
+            } else {
+                state.write(line)?;
+                break;
+            }
+        }
+    }
+    state.require_new_line()?;
+    state.write(TRIPLE_QUOTE)
+}
+
+fn prefer_block_string(value: &str) -> bool {
+    // If a value is multi-line, it likely looks nicer as a block string
+    value.contains('\n')
+}
+
+/// Is it possible to create a serialization that, when fed through
+/// [BlockStringValue](https://spec.graphql.org/October2021/#BlockStringValue()),
+/// returns exactly `value`?
+fn can_be_block_string(value: &str) -> bool {
```

Comment:
This function is missing a check for non-printable characters like here:
https://github.com/graphql/graphql-js/blob/0b7590f0a2b65e6210da2e49be0d8e6c27781af2/src/language/blockString.ts#L66-L79

Do you checking for them in some other place?

Reply:
https://spec.graphql.org/draft/#note-65771

Reply:
This non-normative note is also present [in the October 2021 version](https://spec.graphql.org/October2021/#note-65771) but in that case it is consistent with the grammar:

> [StringCharacter](https://spec.graphql.org/October2021/#StringCharacter)
> * SourceCharacter but not `"` or `\` or LineTerminator
> * `\u` EscapedUnicode
> * `\` EscapedCharacter
>
> [SourceCharacter](https://spec.graphql.org/October2021/#SourceCharacter)
> * U+0009
> * U+000A
> * U+000D
> * U+0020–U+FFFF

In the draft version of the spec [SourceCharacter](https://spec.graphql.org/draft/#SourceCharacter) was changed to “Any Unicode scalar value”, including non-printable ones. I think the note should have been removed at the same time.